From: Sjoerd Simons <sjoerd.simons@collabora.co.uk>
Date: Tue, 28 Jan 2020 16:55:47 +0100
Subject: Correct PID of the atari classic controller

---
 drivers/hid/hid-ids.h | 2 +-
 1 file changed, 1 insertion(+), 1 deletion(-)

diff --git a/drivers/hid/hid-ids.h b/drivers/hid/hid-ids.h
index befbe28..472d3d9 100644
--- a/drivers/hid/hid-ids.h
+++ b/drivers/hid/hid-ids.h
@@ -197,7 +197,7 @@
 
 #define USB_VENDOR_ID_ATARI		0x3250
 #define USB_DEVICE_ID_ATARI_MODERN	0x102b
-#define USB_DEVICE_ID_ATARI_CLASSIC	0x102a
+#define USB_DEVICE_ID_ATARI_CLASSIC	0x101b
 
 #define USB_VENDOR_ID_ATEN		0x0557
 #define USB_DEVICE_ID_ATEN_UC100KM	0x2004
